table/substrait/functions_set.yaml

```diff
@@ -0,0 +1,36 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+%YAML 1.2
+---
+scalar_functions:
+  -
+    name: "is_in"
+    description: >
+      Checks membership of a value in a list of values
+
+      Returns true or false if `needle` is found in `haystack`.
+    impls:
+      - args:
+        - name: needle
+          value: any1
+        - name: haystack
+          value: list<any1>
+        options:
+          nan_equality:
+            values: [ NAN_IS_NAN, NAN_IS_NOT_NAN ]
```

Comment:
I'm not deeply familiar with substrait, so be gentle. Wouldn't it make more sense to rewrite expressions, eg: `a IS IN (null, nan, 'a', 'b', 'c')` to `a IS IN ('a', 'b', 'c') ||  is_nan(a) || not is_null(a)`

Reply:
Just sharing an idea. We could also rewrite the whole thing to just equality expressions :)

Reply:
The reason for this option is because some compute engines recognize `NaN` == `NaN` when performing an `IN` execution, while other compute engines maintain that `NaN != NaN` even in this scenario (and possibly perform the rewrite you're describing when generating the plan). 

In the default substrait extensions there isn't an `is_in` function, there's only [`index_in`](https://github.com/substrait-io/substrait/blob/main/extensions/functions_set.yaml) and I didn't want to implement this as `is_null(index_in(.....))`. But I could do that instead if you prefer

Reply:
Interesting. Thanks for sharing the context around that. I'm okay with the current implementation.

Also, noticed recently that there are more differences: https://github.com/apache/iceberg-python/issues/1295